compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1262,6 +1271,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err.emit();
     }
 
+    fn suggest_ptr_null_mut(
+        &self,
+        expected_ty: Ty<'tcx>,
+        provided_ty: Ty<'tcx>,
+        provided_args: &IndexVec<ProvidedIdx, &hir::Expr<'tcx>>,
+        provided_idx: ProvidedIdx,
+        err: &mut rustc_errors::DiagnosticBuilder<'tcx, ErrorGuaranteed>,
+    ) {
+        if let ty::RawPtr(ty::TypeAndMut { mutbl: expected_mutbl, .. }) = expected_ty.kind()
+            && let ty::RawPtr(ty::TypeAndMut { mutbl: provided_mutbl, .. }) = provided_ty.kind()
+            && let provided_arg_expr = *provided_args[provided_idx]
+            && let hir::ExprKind::Call(callee, _) = provided_arg_expr.kind
+            && let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = callee.kind
+            && let Res::Def(_, def_id) = path.res
+        {
+            match (expected_mutbl, provided_mutbl) {
+                (hir::Mutability::Mut, hir::Mutability::Not) => {
+                    if let Some(ptr_null_def_id) = self.tcx.get_diagnostic_item(sym::ptr_null)
+                        && def_id == ptr_null_def_id
+                    {
+                        // The user provided `ptr::null()`, but the function expects
+                        // `ptr::null_mut()`.
+                        err.subdiagnostic(SuggestPtrNullMut {
+                            span: provided_arg_expr.span
+                        });
+                    }
+                },
+                _ => {},
+            }
+        }
```

Comment:
You can simplify the condition quite a lot, maybe something like

```suggestion
        if let ty::RawPtr(ty::TypeAndMut { mutbl: hir::Mutability::Mut, .. }) = expected_ty.kind()
            && let ty::RawPtr(ty::TypeAndMut { mutbl: hir::Mutability::Not, .. }) = provided_ty.kind()
            && let provided_arg_expr = *provided_args[provided_idx]
            && let hir::ExprKind::Call(callee, _) = provided_arg_expr.kind
            && let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = callee.kind
            && let Res::Def(_, def_id) = path.res
            && Some(def_id) == self.tcx.get_diagnostic_item(sym::ptr_null)
        {
            // The user provided `ptr::null()`, but the function expects
            // `ptr::null_mut()`.
            err.subdiagnostic(SuggestPtrNullMut {
                span: provided_arg_expr.span
            });
        }
```